### ISSUE TITLE
warning against piping to head & tail

### ISSUE THREAD

**Issue by aw-altiscale** [koalaman/shellcheck#429]
On some OSes/certain situations, head and tail may exit early if they have enough data to fulfill the request. This can result in spurious errors from whatever program is feeding them through a pipe.  While not fatal/harmful, it can be confusing.

**Comment by bittorf** [comment]
please explain in detail, how 'head' can fail. 
also please provide an example on which OS/situation tail fails.

**Comment by aw-altiscale** [comment]
This stack overflow question has a much better answer about this particular issue than I could give:
http://superuser.com/questions/554855/how-can-i-fix-a-broken-pipe-error

As you can see, it's hard building a reproducible case.  It's also worth noting that it's the process that feeds the pipe that gets the error, not the head or tail that is reading from the pipe.

In my own code, I've hit it twice in the past year, the last one was a relatively simple construction:

```bash
tail -n "+${startline}" "${file}"  | head -n ${numlines} > "${outfile}"
```

where startline is generally a large number and numlines is a small number.  On OS X w/bash 3.x, it never throws an error.  On Ubunutu w/bash 4.x, I've seen tail throw out the broken pipe write error a few times.

We fixed it in sort of a bad way:

```bash
tail -n "+${startline}" "${file}" 2>/dev/null | head -n ${numlines} > "${outfile}"
```
since throwing errors aware are bad.  (But given the code surrounding this, the risk is primarily if the system is falling over already.... out of mem, etc.)

(The one I hit today which reminded me to file this issue is a little more complex. hah)

**Comment by bittorf** [comment]
very interesting - thank you!

**Comment by aw-altiscale** [comment]
Updated to be just head, since tail ususally doesn't suffer from this problem and one is more likely to see it with head.

**Comment by Arthur2e5** [comment]
`blah | (head; cat >/dev/null)`
